### ISSUE TITLE
Does aria-atomic have real-world use cases?

### ISSUE THREAD

**Issue by aleventhal** [w3c/aria#2458]
Back when ARIA live regions were first developed, aria-atomic seemed like a good idea, in case a tool wanted to update a single node, but the only real way to represent the content was reading a container that it was in.

However, as we improve features on some of our platforms, we're looking to prioritize. A question came up as to whether aria-atomic is used in the real world, or whether it could be deprioritized in ATs, or even deprecated from the ARIA spec.

When I have seen it in markup, it's always been wrong. I see authors throwing it in because, well, that's the general approach to ARIA, aka "if it sounds like it might be right, add it".

Looking for feedback.

**Comment by lolaodelola** [comment]
I think `aria-atomic` can be useful when used properly, do you/the group think lack of (proper) use in the wild is because developers don't find it necessary or because a lack of developer education around the issue? Could AT implementation also be affecting how it's used? i.e. is implementation consistent across ATs?

I think in cases where a status or error/notice message may need to be updated `aria-atomic` comes in very handy.

**Comment by lolaodelola** [comment]
I haven't used it myself but off the top of my head, an example I'm thinking of is a log in form that dynamically updates it's notices or error message. E.g. a user puts in the wrong email in a log in form and the text "email required" changes to an error "invalid email"

**Comment by aleventhal** [comment]
That should work without aria-atomic, afaict

**Comment by spectranaut** [comment]
In today's ARIA meeting, we decided to make this attribute deprecated -- since aria-atomic does not have consistent support among browsers, fine grained live region announcements can be handled by a hidden live region, and aria-notify support is coming soon.

Minutes from meeting: https://www.w3.org/2025/03/06-aria-minutes#9ee4

I'll make the PR.

**Comment by spectranaut** [comment]
I think that I will wait on this PR until ariaNotify lands in the spec, so the deprecation note can refer to ariaNotify.

**Comment by yatil** [comment]
Oof. I frequently use and recommend the use of aria-atomic to avoid duplication of notification logic.

My prime example would be something like


```html
<button aria-live="polite" aria-atomic="true">Cart. <span id="items">3</span> <span class="visually-hidden">Items</span></button>
```

Changing `#items` creates a nice status message. Without `aria-atomic="true"` only the changed number would be announced.

Conversely I had the use case of changing a part of the message where announcing only that part was sufficient, so the `aria-atomic="false"` behavior was welcome.

I would rather not give up that flexibility.

**Comment by yatil** [comment]
Hidden live regions – like all hidden content – should be a last resort. The accessibility bugs I see on a daily basis that result of content being hidden but overlooked is ridiculous.

**Comment by patrickhlauke** [comment]
I've had instances in the past (would need to dig around/retest) where injecting content into a live region that was previously announced already would result in just the difference being announced, rather than the whole thing, even though the entire content of the region was effectively replaced. some "clever" heuristic perhaps that determined what the delta was. in those cases, I had to resort to forcing the entire thing to be re-announced with `aria-atomic="true"`, or first clearing/emptying the live region for a brief moment before then reinjecting the content.

**Comment by MarcoZehe** [comment]
When I was contributing to the Element/Matrix web client, we ran into an issue where we had lots of duplicated speech when new chat messages arrived, reactions were added to previously posted messages etc. The way to solve this and avoid overcealous speech was by using aria-atomic properly. This then worked fine in Firefox and Chrome, including the desktop app, with both VoiceOver on the Mac and NVDA on Windows. Deprecating this would be detrimental to providing good accessibility for those cases. The example @yatil gave is also a very valid example where extraneous speech is being avoided. So, I am strictly against deprecating it and keeping it in the spec proper.

**Comment by yatil** [comment]
I agree that my use cases and tests with screen reader and browser implementations have never shown any significant lack of support. The meeting minutes – and hence the resolution here – seem to be based on anecdata. The [latest tests on a11ysupport.io](https://a11ysupport.io/tech/aria/aria-atomic_attribute) are two years old and score 37/44. And the open issues seem relatively simple to overcome for browser makers.

I agree that the live-region-based messaging might be suboptimal for many use cases, and if a new technology can be better for those, then that’s great. But declarative ways to specify and control how messages are announced are important, too. This is especially necessary for those who don’t see the web primarily as a JavaScript platform.

**Comment by pkra** [comment]
@patrickhlauke  the current process is documented here: https://github.com/w3c/aria/blob/main/documentation/process.md  Before that, a commitment to implement was sufficient.

**Comment by scottaohara** [comment]
I respect the initial reactions to the idea of deprecating this feature.  I'm not 100% sold on deprecation... but I wouldn't object to it.  Specifically because:

1. @spectranaut mentioned the spec wouldn't be updated until ariaNotify landed.
2. deprecation isn't the same as removal - so the use cases being mentioned wouldn't stop working. But ariaNotify would be the recommended alternative to continued aria-atomic use.  Sure, deprecation means bugs with the attribute's behavior may not get resolved... but that seems like the current reality?
3. `aria-atomic` and ariaNotify are/will both reliant on JS to work (that live region ain't going to repopulate itself). Deprecating aria-atomic in favor of a new JS solution that decouples the need for authors to define some functionality in JS and some functionality in the markup to get the desired output seems kinda ok to me.

**Comment by yatil** [comment]
Can someone link to that mysterious ariaNotify specification? And maybe outline how that can be a drop-in replacement for aria-atomic?

Of course you need JS for aria-atomic, but live regions (sometimes using aria-atomic) allow re-use of the available UI elements instead of re-implementing additional logic for messages. Especially as people often underestimate screen reader users and deliver overly verbose messages.

**Comment by andrewnordlund** [comment]
Deprecating this is going to break a lot of legacy code that was properly written.  Now everyone who may have ever written anything with `aria-atomic` is going to have to go through all their legacy code looking for instances of it, and then re-write it in hopes that `aria-notify` will be better supported.

Gee.  I wonder why it's so hard to get developers, and the companies that employ them, to write accessible websites.

**Comment by andrewnordlund** [comment]
A few years ago I got frustrated with inaccessible password-requirements on websites.  So I wrote an accessible password-requirements widget:  https://github.com/andrewnordlund/passwordRequirements

I have since left the field of accessibility as a profession, in part because of moves like this.  Accessibility is hard enough without constantly moving the goalposts like this.  If I have to re-write that widget, I no longer have access to all the AT I used to in order to test it to see if it works in different stacks.

**Comment by yatil** [comment]
Just clarifying: depreciation in W3C parlance usually means "please don't use this in new code", it's rare that the existing implementations cease to work.

**Comment by aleventhal** [comment]
I agree we should not deprecate now that we have real examples. That's all we needed in order to keep the feature.
In particular yatil's shopping cart example is great.
Thank you everyone for the input!

**Comment by aleventhal** [comment]
@spectranaut I know we had consensus but now that we have new info I move that we close this issue.

**Comment by rahimabdi** [comment]
@yatil Here's a link to the `ariaNotify` explainer: https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Accessibility/AriaNotify/explainer.md.

**Comment by LJWatson** [comment]
I think it would be a mistake to deprecate `aria-atomic`. When implemented correctly it serves the original use case well, and with far less code than hidden live regions would require - even if the concerns about hidden code already mentioned weren't a thing - which they very much are.

If improper use was a good reason for deprecation, then I fear the entire ARIA spec would need to be deprecated.
